src/responder/pam/pamsrv_passkey.c

```diff
@@ -0,0 +1,923 @@
+
+        ret = pam_add_response(pctx->pd, SSS_PAM_PASSKEY_INFO, strlen(prompt_pin) + 1,
+                               (const uint8_t *) prompt_pin);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_CRIT_FAILURE, "pam_add_response failed. [%d]: %s\n",
+                  ret, sss_strerror(ret));
+            goto done;
+        }
+
+        pctx->pd->pam_status = PAM_SUCCESS;
+        pam_reply(pctx->preq);
+        talloc_free(pk_data);
+        return;
+    }
+
+    req = pam_passkey_auth_send(pctx, pctx->ev, timeout, debug_libfido2,
+                                verification, pctx->pd, pk_data);
+    if (req == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "pam_passkey_auth_send failed [%d]: %s\n",
+                                 ret, sss_strerror(ret));
+        goto done;
+    }
+
+    tevent_req_set_callback(req, pam_forwarder_passkey_cb, pctx->preq);
+
+done:
+    if (pk_data != NULL) {
+        talloc_free(pk_data);
+    }
+
+    if (ret == ENOENT) {
+        /* No passkey data, continue through to typical auth flow */
+        DEBUG(SSSDBG_TRACE_FUNC, "No passkey data found, skipping passkey auth\n");
+        pam_check_user_search(pctx->preq);
+    } else if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Unexpected passkey error [%d]: %s."
+                                 " Skipping passkey auth\n",
+                                 ret, sss_strerror(ret));
+        pam_check_user_search(pctx->preq);
+    }
+
+    return;
+}
+
+
+struct pam_passkey_auth_send_state {
+    struct pam_data *pd;
+    struct tevent_context *ev;
+    struct sss_child_ctx_old *child_ctx;
+    struct child_io_fds *io;
+    const char *logfile;
+    const char **extra_args;
+    char *verify_opts;
+    int timeout;
+    int child_status;
+};
+
+static errno_t passkey_child_exec(struct tevent_req *req);
+static void pam_passkey_auth_done(int child_status,
+                                  struct tevent_signal *sige,
+                                  void *pvt);
+
+errno_t get_passkey_child_write_buffer(TALLOC_CTX *mem_ctx,
+                                       struct pam_data *pd,
+                                       uint8_t **_buf, size_t *_len)
+{
+    int ret;
+    uint8_t *buf;
+    size_t len;
+    const char *pin = NULL;
+
+    if (pd == NULL || pd->authtok == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing authtok.\n");
+        return EINVAL;
+    }
+
+    if (sss_authtok_get_type(pd->authtok) == SSS_AUTHTOK_TYPE_PASSKEY) {
+        ret = sss_authtok_get_passkey_pin(pd->authtok, &pin, &len);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE, "sss_authtok_get_sc_pin failed [%d]: %s\n",
+                                    ret, sss_strerror(ret));
+            return ret;
+        }
+        if (pin == NULL || len == 0) {
+            DEBUG(SSSDBG_OP_FAILURE, "Missing PIN.\n");
+            return EINVAL;
+        }
+
+        buf = talloc_size(mem_ctx, len);
```

Comment:
Hi,

I would suggest to add `sss_erase_talloc_mem_securely()` as a talloc destructor to `buf` to make sure the pin is properly overwritten.

Additionally, the PIN is send without `\0` in the end which is completely ok. I found during testing that `passkey_child` might need some fix to make sure that it will not be confused about the length:
```
diff --git a/src/passkey_child/passkey_child_credentials.c b/src/passkey_child/passkey_child_credentials.c
index 9ebab97db..1eaf7bcc3 100644
--- a/src/passkey_child/passkey_child_credentials.c
+++ b/src/passkey_child/passkey_child_credentials.c
@@ -173,7 +173,7 @@ passkey_recv_pin(TALLOC_CTX *mem_ctx, int fd, char **_pin)
         return EINVAL;
     }
 
-    str = talloc_strdup(mem_ctx, (char *) buf);
+    str = talloc_strndup(mem_ctx, (char *) buf, len);
     if (str == NULL) {
         DEBUG(SSSDBG_CRIT_FAILURE, "talloc_strndup failed.\n");
         return ENOMEM;
```
`buf` is just a fixed size `uint8_t` array which is filled with the PIN. Since `buf` might contain random data the next byte following the received PIN might not be `\0` and as a result more bytes than expected will be copied into `str`. Using `talloc_strndup()` fixes this. I think it would be most straight forward to add this or a similar fix into your PR.

bye,
Sumit

Reply:
Thanks Sumit. I added a commit which adds the destructor for the PIN as you stated.

The other issue is already addressed in 7fb89ab01312bba9cb4099dc874725801f12bc78 if I am not wrong, I rebased this PR so it should be included now.